### ISSUE TITLE
Error with MSFRPCd

### ISSUE THREAD

**Issue by voidnecron** [SerpicoProject/Serpico#424]
Serpico's MSFRPCd seems to be faulting in the combination of latest Kali (kali-linux-2018.1) and latest Serpico (1.2.2+20180104221549).
Manual for setting up the connection which was followed: https://github.com/SerpicoProject/Serpico/wiki/Setting-up-Metasploit-RPC-Connection

1. Follow instructions from manual to establish connection (see URL above)
2. Go to a report and press 'hosts' for instance
3. *crash*
Error at console:
https://10.0.2.15:8443/report/1/additional_features -> /report/1/msfsettings
2018-02-16 08:51:47 - Msf::RPC::Exception - Client not authenticated:
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/msfrpc-client-1.1.1/lib/msfrpc-client/client.rb:111:in `call'
	/opt/Serpico/Serpico/routes/report.rb:1866:in `block in <top (required)>'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1632:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1632:in `block in compile!'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:991:in `block (3 levels) in route!'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1010:in `route_eval'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:991:in `block (2 levels) in route!'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1037:in `block in process_route'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1035:in `catch'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1035:in `process_route'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:989:in `block in route!'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:988:in `each'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:988:in `route!'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1001:in `route!'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1094:in `block in dispatch!'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1073:in `block in invoke'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1073:in `catch'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1073:in `invoke'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1091:in `dispatch!'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:923:in `block in call!'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1073:in `block in invoke'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1073:in `catch'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1073:in `invoke'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:923:in `call!'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:913:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/rack-protection-2.0.0/lib/rack/protection/xss_header.rb:18:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/rack-protection-2.0.0/lib/rack/protection/base.rb:50:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/rack-protection-2.0.0/lib/rack/protection/base.rb:50:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/rack-protection-2.0.0/lib/rack/protection/path_traversal.rb:16:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/rack-protection-2.0.0/lib/rack/protection/json_csrf.rb:26:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/rack-protection-2.0.0/lib/rack/protection/base.rb:50:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/rack-protection-2.0.0/lib/rack/protection/base.rb:50:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/rack-protection-2.0.0/lib/rack/protection/frame_options.rb:31:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/rack-2.0.3/lib/rack/session/abstract/id.rb:232:in `context'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/rack-2.0.3/lib/rack/session/abstract/id.rb:226:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/rack-2.0.3/lib/rack/logger.rb:15:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/rack-2.0.3/lib/rack/common_logger.rb:33:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:231:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:224:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/rack-2.0.3/lib/rack/head.rb:12:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/rack-2.0.3/lib/rack/method_override.rb:22:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/show_exceptions.rb:22:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:194:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1955:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1499:in `block in call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1726:in `synchronize'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/sinatra-2.0.0/lib/sinatra/base.rb:1499:in `call'
	/opt/Serpico/embedded/lib/ruby/gems/2.3.0/gems/rack-2.0.3/lib/rack/handler/webrick.rb:86:in `service'
	/opt/Serpico/embedded/lib/ruby/2.3.0/webrick/httpserver.rb:140:in `service'
	/opt/Serpico/embedded/lib/ruby/2.3.0/webrick/httpserver.rb:96:in `run'
	/opt/Serpico/embedded/lib/ruby/2.3.0/webrick/server.rb:296:in `block in start_thread'

**Comment by mart787** [comment]
I've been struggling with this. I managed to get it to work again by downgrading Serpico to 1.2.1

**Comment by melcher83** [comment]
@mart7I'm having the same issue. How do you downgrade?

**Comment by mart787** [comment]
I used dpkg to remove and then installed the older version from github. Kept all the data and worked first time..

**Comment by Petessss** [comment]
What version of Metasploit are you using?

**Comment by mart787** [comment]
latest one on kali 2018 via apt-get

4.16.39-dev

**Comment by voidnecron** [comment]
@Petessss I was asked to create this issue by @BuffaloWill. He thought he might have wrecked something with the latest update since he was fiddling around with the cyphers.

**Comment by voidnecron** [comment]
1.2.1 doesn't have the issue, so that confirms it's something in the latest release of Serpico and not something metasploit might have done.

**Comment by Petessss** [comment]
Cool. I wanted to make  sure you didn't think the ticket went into a black hole. :)

**Comment by SchellerCyberSecurity** [comment]
I downgraded to version 1.2.1 and I am still having this issue.

**Comment by BuffaloWill** [comment]
We need to get odle merged in or push a fix for MSFRPC in the mean time.

@ccammilleri any idea how difficult of a fix this is?

**Comment by agreenbhm** [comment]
I believe I was able to figure out a workaround.  I compared the changes between release 1.2.1 and 1.2.2 (since that seems to be where things broke), and it seems that the dependency "msfrpc-client" was upgraded from version 1.0.3 to 1.1.1.  To fix, I modified Gemfile and Gemfile.lock and changed all references to msfrpc-client to show version 1.0.3. This has two dependencies of it's own (msgpack and librex) which were missing afterwards, but bundle gave me a helpful message that I could just run "bundle install msfrpc-client" to install the dependencies plus the needed version of the msf library.  I started Serpico up after that, pointed it at my msfrpcd, and I was able to import hosts.

So, in short:
1. Edit Gemfile and Gemfile.lock to change the version of msfrpc-client to 1.0.3
2. Install updated Gemfile by running "bundle install" (which will fail, but I'm not sure if this step was necessary for some other parts, first.  Did I mention I know next to nothing about Ruby?)
3. Install the corrected version of msfrpc-client by running "bundle update msfrpc-client" 
4. Start Serpico
5. Configure Serpico's msfrpcd settings like normal
6. Import hosts

Let me know if that works/doesn't work for you.

**Comment by agreenbhm** [comment]
I have created a pull request to merge my changes into the master branch: https://github.com/SerpicoProject/Serpico/pull/519

**Comment by noraj** [comment]
On ruby 2.6.3.

```
gem install msfrpc-client -v 1.0.3
Fetching msfrpc-client-1.0.3.gem
Fetching librex-0.0.999.gem
Fetching msgpack-0.5.12.gem
Building native extensions. This could take a while...
ERROR:  Error installing msfrpc-client:
        ERROR: Failed to build gem native extension.

    current directory: /home/noraj/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/msgpack-0.5.12/ext/msgpack
/home/noraj/.rbenv/versions/2.6.3/bin/ruby -I /home/noraj/.rbenv/versions/2.6.3/lib/ruby/2.6.0 -r ./siteconf20190606-25288-1vdrjl0.rb extconf.rb
checking for ruby/st.h... yes
checking for st.h... yes
checking for rb_str_replace() in ruby.h... yes
checking for rb_intern_str() in ruby.h... yes
checking for rb_sym2str() in ruby.h... yes
checking for rb_str_intern() in ruby.h... yes
creating Makefile

current directory: /home/noraj/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/msgpack-0.5.12/ext/msgpack
make "DESTDIR=" clean

current directory: /home/noraj/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/msgpack-0.5.12/ext/msgpack
make "DESTDIR="
compiling buffer.c
compiling buffer_class.c
compiling core_ext.c
core_ext.c: In function ‘MessagePack_core_ext_module_init’:
core_ext.c:121:22: error: ‘rb_cFixnum’ undeclared (first use in this function); did you mean ‘rb_isalnum’?
     rb_define_method(rb_cFixnum, "to_msgpack", Fixnum_to_msgpack, -1);
                      ^~~~~~~~~~
                      rb_isalnum
core_ext.c:121:22: note: each undeclared identifier is reported only once for each function it appears in
core_ext.c:122:22: error: ‘rb_cBignum’ undeclared (first use in this function); did you mean ‘rb_isalnum’?
     rb_define_method(rb_cBignum, "to_msgpack", Bignum_to_msgpack, -1);
                      ^~~~~~~~~~
                      rb_isalnum
core_ext.c: At top level:
cc1: warning: unrecognized command line option ‘-Wno-self-assign’
cc1: warning: unrecognized command line option ‘-Wno-parentheses-equality’
cc1: warning: unrecognized command line option ‘-Wno-constant-logical-operand’
make: *** [Makefile:244: core_ext.o] Error 1

make failed, exit code 2

Gem files will remain installed in /home/noraj/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/msgpack-0.5.12 for inspection.
Results logged to /home/noraj/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/extensions/x86_64-linux/2.6.0-static/msgpack-0.5.12/gem_make.out
```

Please update ruby to 2.6.x branch #563

**Comment by noraj** [comment]
With ruby 2.6.3 if I let bundle install do I get Installing msfrpc-client 1.1.1 and it works.

**Comment by BuffaloWill** [comment]
Landed.
